### ISSUE TITLE
[using taskchampion] Task end set to "" when marked completed

### ISSUE THREAD

**Issue by djmitche** [GothenburgBitFactory/taskwarrior#3109]
This is a task sync'd via bugwarrior. Its `end` property is currently an empty string, which Taskwarrior is interpreting as the epoch (1970).

```
sqlite> select * from tasks where uuid = 'bff35d18-c803-4c15-bb16-f6377a106845';
bff35d18-c803-4c15-bb16-f6377a106845|{"start":"1685022931","tags_gerrit":"x","gerritid":"4558705","tags":"gerrit,inprogress","priority":"M","gerritbranch":"main","end":"","modified":"1684879333","wait":"1684949097","entry":"1684879333","gerrittopic":"notopic","description":"https://crrev.com/c/4558705 Add the Sec-CH-UA-Form-Factor header","gerritsummary":"Add the Sec-CH-UA-Form-Factor header","status":"pending","tags_inprogress":"x","gerriturl":"https://chromium-review.googlesource.com/#/c/4558705/"}
sqlite> select * from operations where data like '%bff35d18-c803-4c15-bb16-f6377a106845%';
8075|{"Create":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845"}}
8076|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"description","old_value":null,"value":"https://crrev.com/c/4558705 Add the Sec-CH-UA-Form-Factor header","timestamp":"2023-05-23T22:02:13.394701172Z"}}
8077|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"entry","old_value":null,"value":"1684879333","timestamp":"2023-05-23T22:02:13.400090116Z"}}
8078|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"gerritbranch","old_value":null,"value":"main","timestamp":"2023-05-23T22:02:13.407460750Z"}}
8079|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"gerritid","old_value":null,"value":"4558705","timestamp":"2023-05-23T22:02:13.413536166Z"}}
8080|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"gerritsummary","old_value":null,"value":"Add the Sec-CH-UA-Form-Factor header","timestamp":"2023-05-23T22:02:13.418142289Z"}}
8081|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"gerrittopic","old_value":null,"value":"notopic","timestamp":"2023-05-23T22:02:13.422746678Z"}}
8082|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"gerriturl","old_value":null,"value":"https://chromium-review.googlesource.com/#/c/4558705/","timestamp":"2023-05-23T22:02:13.428777777Z"}}
8083|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"modified","old_value":null,"value":"1684879333","timestamp":"2023-05-23T22:02:13.434230583Z"}}
8084|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"priority","old_value":null,"value":"M","timestamp":"2023-05-23T22:02:13.438549296Z"}}
8085|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"status","old_value":null,"value":"pending","timestamp":"2023-05-23T22:02:13.448630389Z"}}
8086|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"tags","old_value":null,"value":"gerrit,inprogress","timestamp":"2023-05-23T22:02:13.454071139Z"}}
8087|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"tags_gerrit","old_value":null,"value":"x","timestamp":"2023-05-23T22:02:13.458476808Z"}}
8088|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"tags_inprogress","old_value":null,"value":"x","timestamp":"2023-05-23T22:02:13.462803668Z"}}
8121|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"start","old_value":null,"value":"1684882237","timestamp":"2023-05-23T22:50:37.866844275Z"}}
8145|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"start","old_value":"1684882237","value":null,"timestamp":"2023-05-24T13:23:23.963750165Z"}}
8169|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"wait","old_value":null,"value":"1684949097","timestamp":"2023-05-24T15:24:57.540511221Z"}}
8227|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"end","old_value":null,"value":"1684954933","timestamp":"2023-05-24T19:02:13.234340594Z"}}
8228|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"status","old_value":"pending","value":"completed","timestamp":"2023-05-24T19:02:13.239252334Z"}}
8262|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"end","old_value":"1684954933","value":"","timestamp":"2023-05-24T21:02:13.450598400Z"}}
8263|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"status","old_value":"completed","value":"pending","timestamp":"2023-05-24T21:02:13.455741993Z"}}
8317|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"start","old_value":null,"value":"1685022908","timestamp":"2023-05-25T13:55:08.891262516Z"}}
8319|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"start","old_value":"1685022908","value":null,"timestamp":"2023-05-25T13:55:25.723095994Z"}}
8321|{"Update":{"uuid":"bff35d18-c803-4c15-bb16-f6377a106845","property":"start","old_value":null,"value":"1685022931","timestamp":"2023-05-25T13:55:31.070263127Z"}}
```

So at 2023-05-24T19:02:13.234340594Z its `end` property was first set to 1684954933, then its `status` set to "completed", then its `end` set to an empty string. It was then re-opened at 2023-05-24T21:02:13.455741993Z and appears now as
```
⸩ task 37

Name             Value                                                           
ID               37                                                              
Description      https://crrev.com/c/4558705 Add the Sec-CH-UA-Form-Factor header
...
End              1970-01-01 00:00:00                                             
```

**Comment by djmitche** [comment]
I can't reproduce this by just marking a task with `task done` and then re-starting it.  I need to figure out what bugwarrior is doing "under the hood".

**Comment by djmitche** [comment]
Here's an strace from re-opening #3102:
`[pid 841589] execve("/home/dustin/bin/task", ["task", "rc.verbose=new-uuid", "rc.json.array=TRUE", "rc.confirmation=no", "rc.dependency.confirmation=no", "rc.recurrence.confirmation=no", "rc.uda.githubtitle.type=string", "rc.uda.githubtitle.label=\"Github"..., "rc.uda.githubbody.type=string", "rc.uda.githubbody.label=\"Github "..., "rc.uda.githubcreatedon.type=date", "rc.uda.githubcreatedon.label=\"Gi"..., "rc.uda.githubupdatedat.type=date", "rc.uda.githubupdatedat.label=\"Gi"..., "rc.uda.githubclosedon.type=date", "rc.uda.githubclosedon.label=\"Git"..., "rc.uda.githubmilestone.type=stri"..., "rc.uda.githubmilestone.label=\"Gi"..., "rc.uda.githubrepo.type=string", "rc.uda.githubrepo.label=\"Github "..., "rc.uda.githuburl.type=string", "rc.uda.githuburl.label=\"Github U"..., "rc.uda.githubtype.type=string", "rc.uda.githubtype.label=\"Github "..., "rc.uda.githubnumber.type=numeric", "rc.uda.githubnumber.label=\"Githu"..., "rc.uda.githubuser.type=string", "rc.uda.githubuser.label=\"Github "..., "rc.uda.githubnamespace.type=stri"..., "rc.uda.githubnamespace.label=\"Gi"..., "rc.uda.githubstate.type=string", "rc.uda.githubstate.label=\"GitHub"..., "514efd17-5a31-4dee-9bf3-6279f08f"..., "modify", "end:", "githubupdatedat:\"20230528T200805"..., "status:\"pending\""], ["SHELL=/bin/bash", "HISTCONTROL=erasedups", "TERM_PROGRAM_VERSION=3.2a", "TMUX=/tmp/tmux-1001/default,7931"..., "HISTSIZE=2000", "SSH_AUTH_SOCK=/tmp/dorp-1001/age"..., "EDITOR=vim", "PWD=/home/dustin", "LOGNAME=dustin", "XDG_SESSION_TYPE=tty", "MOTD_SHOWN=pam", "HOME=/home/dustin", "LANG=en_US.UTF-8", "SSH_CONNECTION=37.19.221.171 386"..., "XDG_SESSION_CLASS=user", "TERM=screen", "USER=dustin", "TMUX_PANE=%0", "SHLVL=2", "XDG_SESSION_ID=7079", "XDG_RUNTIME_DIR=/run/user/1001", "SSH_CLIENT=37.19.221.171 38664 2"..., "XDG_DATA_DIRS=/usr/local/share:/"..., "PATH=/home/dustin/bin:/usr/local"..., "HISTIGNORE=ls:cd", "DBUS_SESSION_BUS_ADDRESS=unix:pa"..., "PASSWORD_STORE_DIR=/home/dustin/"..., "SSH_TTY=/dev/pts/0", "TERM_PROGRAM=tmux", "_=/usr/bin/strace", "TASKRC=/home/dustin/.taskrc"]) = 0`

The meat of that is `["task", .., "514efd17-5a31-4dee-9bf3-6279f08f"..., "modify", "end:", "githubupdatedat:\"20230528T200805"..., "status:\"pending\""]` so in particular it's explicitly setting `end:`,

**Comment by djmitche** [comment]
OK, and I can reproduce on the CLI then:

```
✺ funk ~/p/taskwarrior ⸨working-branch⸩
⸩ task add test
TASKRC override: /home/dustin/p/taskwarrior/.taskrc
Created task 1.
✺ funk ~/p/taskwarrior ⸨working-branch⸩
⸩ task 1 done
Completed task 1 'test'.
Completed 1 task.
TASKRC override: /home/dustin/p/taskwarrior/.taskrc
✺ funk ~/p/taskwarrior ⸨working-branch⸩
⸩ task 1 modify end: status:pending
Modifying task 1 'test'.
Modified 1 task.
TASKRC override: /home/dustin/p/taskwarrior/.taskrc
✺ funk ~/p/taskwarrior ⸨working-branch⸩
⸩ task 1
TASKRC override: /home/dustin/p/taskwarrior/.taskrc

Name          Value                               
ID            1                                   
Description   test
Status        Pending                             
Entered       2023-05-28 20:19:50 (13s)
End           1970-01-01 00:00:00                 
Last modified 2023-05-28 20:20:01 (2s)
Virtual tags  LATEST PENDING READY UNBLOCKED      
UUID          3735d6e2-38a9-46d5-af2c-c5a6cc5bd356
Urgency       0                                   

✺ funk ~/p/taskwarrior ⸨working-branch⸩
⸩ sqlite3 ~/.task/taskchampion.sqlite3 .dump
PRAGMA foreign_keys=OFF;
BEGIN TRANSACTION;
CREATE TABLE operations (id INTEGER PRIMARY KEY AUTOINCREMENT, data STRING);
INSERT INTO operations VALUES(1,'"UndoPoint"');
INSERT INTO operations VALUES(2,'{"Create":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356"}}');
INSERT INTO operations VALUES(3,'{"Update":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356","property":"description","old_value":null,"value":"test","timestamp":"2023-05-28T20:19:50.066388405Z"}}');
INSERT INTO operations VALUES(4,'{"Update":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356","property":"entry","old_value":null,"value":"1685305190","timestamp":"2023-05-28T20:19:50.076265749Z"}}');
INSERT INTO operations VALUES(5,'{"Update":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356","property":"modified","old_value":null,"value":"1685305190","timestamp":"2023-05-28T20:19:50.083100605Z"}}');
INSERT INTO operations VALUES(6,'{"Update":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356","property":"status","old_value":null,"value":"pending","timestamp":"2023-05-28T20:19:50.098394824Z"}}');
INSERT INTO operations VALUES(7,'"UndoPoint"');
INSERT INTO operations VALUES(8,'{"Update":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356","property":"end","old_value":null,"value":"1685305193","timestamp":"2023-05-28T20:19:53.890176127Z"}}');
INSERT INTO operations VALUES(9,'{"Update":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356","property":"modified","old_value":"1685305190","value":"1685305193","timestamp":"2023-05-28T20:19:53.898562789Z"}}');
INSERT INTO operations VALUES(10,'{"Update":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356","property":"status","old_value":"pending","value":"completed","timestamp":"2023-05-28T20:19:53.907746159Z"}}');
INSERT INTO operations VALUES(11,'"UndoPoint"');
INSERT INTO operations VALUES(12,'{"Update":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356","property":"end","old_value":"1685305193","value":"","timestamp":"2023-05-28T20:20:01.171453608Z"}}');
INSERT INTO operations VALUES(13,'{"Update":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356","property":"modified","old_value":"1685305193","value":"1685305201","timestamp":"2023-05-28T20:20:01.180137093Z"}}');
INSERT INTO operations VALUES(14,'{"Update":{"uuid":"3735d6e2-38a9-46d5-af2c-c5a6cc5bd356","property":"status","old_value":"completed","value":"pending","timestamp":"2023-05-28T20:20:01.186455105Z"}}');
CREATE TABLE sync_meta (key STRING PRIMARY KEY, value STRING);
CREATE TABLE tasks (uuid STRING PRIMARY KEY, data STRING);
INSERT INTO tasks VALUES('3735d6e2-38a9-46d5-af2c-c5a6cc5bd356','{"modified":"1685305201","entry":"1685305190","end":"","status":"pending","description":"test"}');
CREATE TABLE working_set (id INTEGER PRIMARY KEY, uuid STRING);
INSERT INTO working_set VALUES(1,'3735d6e2-38a9-46d5-af2c-c5a6cc5bd356');
DELETE FROM sqlite_sequence;
INSERT INTO sqlite_sequence VALUES('operations',14);
COMMIT;
```

**Comment by djmitche** [comment]
Repeating the same thing with the current released version:
```
[description:"test" entry:"1685305353" modified:"1685305366" priority:"M" status:"pending" uuid:"92d3c0f4-8eb9-47de-8bd6-54776461d84b"]
```
and in `undo.log`:
```
---
time 1685305353
new [description:"test" entry:"1685305353" modified:"1685305353" priority:"M" status:"pending" uuid:"92d3c0f4-8eb9-47de-8bd6-54776461d84b"]
---
time 1685305358
old [description:"test" entry:"1685305353" modified:"1685305353" priority:"M" status:"pending" uuid:"92d3c0f4-8eb9-47de-8bd6-54776461d84b"]
new [description:"test" end:"1685305358" entry:"1685305353" modified:"1685305358" priority:"M" status:"completed" uuid:"92d3c0f4-8eb9-47de-8bd6-54776461d84b"]
---
time 1685305366
old [description:"test" end:"1685305358" entry:"1685305353" modified:"1685305358" priority:"M" status:"completed" uuid:"92d3c0f4-8eb9-47de-8bd6-54776461d84b"]
new [description:"test" entry:"1685305353" modified:"1685305366" priority:"M" status:"pending" uuid:"92d3c0f4-8eb9-47de-8bd6-54776461d84b"]
---
```
so it seems that `end:` is being treated as deleting that property, so we should probably do the same.

However, in order to adhere to the "any task is valid" aspiration of Taskchampion, we should also accept a task with an empty string for its `end` property, and represent that as if the property did not exist.

**Comment by djmitche** [comment]
Hm, `ColumnTypeDate::validate` already says this is an invalid value. And `Task::get_date` returns the same for a missing property and a property that does not parse as an integer. So this might be as good as we can get for "accepting a task".

**Comment by djmitche** [comment]
Fixed in #3094 (which is where the bug was introduced).
